creator-node/src/services/PrometheusMetric.js

```diff
@@ -0,0 +1,143 @@
+const client = require('prom-client')
+
+const PrometheusType = Object.freeze({
+  HISTOGRAM: 'histogram',
+  GAUGE: 'gauge'
+})
+
+/**
+ * Class to maintain library-specific logic
+ * for Prometheus-bound metrics
+ */
+class PrometheusMetric {
+  // use these static vars to avoid name collisions
+  // and register collectors to be run by the PrometheusMetricExporter
+  static histograms = {}
+  static gauges = {}
+  static registeredCollectors = {}
+
+  /**
+   * Helper to avoid naming collisions.
+   */
+  _initMetric(
+    name,
+    description,
+    labelNames,
+    collection,
+    PrometheusMetricClass
+  ) {
+    if (!(name in collection)) {
+      collection[name] = new PrometheusMetricClass({
+        name: name,
+        help: description,
+        labelNames: labelNames
+      })
+    }
+    this.metric = collection[name]
+  }
+
+  constructor(name, description, labelNames, metric_type) {
+    // set this.startTime to now()
+    this.resetTimer()
+
+    // set metric prefix of audius_project_
+    name = `audius_cn_${name}`
+
+    this.metricType = metric_type
+    if (this.metricType === undefined) {
+      this.metricType = PrometheusType.HISTOGRAM
+    }
+
+    // CollectorRegistries must be uniquely named
+    // NOTE: we only set labelNames once.
+    // unsure if overloading is supported.
+    if (this.metricType === PrometheusType.HISTOGRAM) {
+      this._initMetric(
+        name,
+        description,
+        labelNames,
+        PrometheusMetric.histograms,
+        client.Histogram
+      )
+    } else if (this.metricType === PrometheusType.GAUGE) {
+      this._initMetric(
+        name,
+        description,
+        labelNames,
+        PrometheusMetric.gauges,
+        client.Gauge
+      )
+    } else {
+      throw new Error(`metric_type '${this.metricType}' not found`)
+    }
+  }
+
+  /**
+   * Useful when timing complex functionality.
+   */
+  resetTimer() {
+    this.startTime = Date.now()
+  }
+
+  /**
+   * Return elapsed time since metric was created.
+   * Useful for logs.
+   * @param {number} [startTime=this.startTime] Return elapsed time since startTime.
+   */
+  elapsed(startTime) {
+    if (startTime === undefined) {
+      startTime = this.startTime
+    }
+    return Date.now() - startTime
+  }
+
+  /**
+   * Save a metric with the value of elapsed time since a start time.
+   * @param {Object[]} [labels] An array of label:value pairs to apply to the metric.
+   * @param {number} [startTime] The startTime to use with this data point.
+   */
+  saveTime(labels, startTime) {
+    this.save(this.elapsed(startTime), labels)
+  }
```

Comment:
`saveTime` is saving the time it took between when the constructor was called (first two lines of the constructor) and when `saveTime()` is called.

However, you can choose to overload saveTime for different reasons while still relying on the same metric name.

What startTime can't be is `Date.now()` since we'll want `Date.now() - startTime` as seen in `elapsed()`.